File: core/src/com/qtech/bubbleblaster/init/GameTypes.java
package com.qtech.bubbleblaster.init;

import com.qtech.bubbleblaster.QInternalAddon;
import com.qtech.bubbleblaster.common.gametype.AbstractGameType;
import com.qtech.bubbleblaster.common.init.ObjectInit;
import com.qtech.bubbleblaster.gametype.ClassicType;
import com.qtech.bubbleblaster.registry.DeferredRegister;
import com.qtech.bubbleblaster.registry.Registers;
import com.qtech.bubbleblaster.registry.object.RegistryObject;

import java.util.function.Supplier;

/**
 * @author Quinten Jungblut
 * @see AbstractGameType
 * @since 1.0.0
 */
@SuppressWarnings("unused")
//@ObjectHolder(addonId = "qbubbles", type = GameType.class)
public class GameTypes implements ObjectInit<AbstractGameType> {
    public static final DeferredRegister<AbstractGameType> GAME_TYPES = DeferredRegister.create(QInternalAddon.ADDON_ID, Registers.GAME_TYPES);

    public static final RegistryObject<ClassicType> CLASSIC_TYPE = register("classic", () -> new ClassicType());

    @SuppressWarnings("SameParameterValue")
    private static <T extends AbstractGameType> RegistryObject<T> register(String name, Supplier<T> supplier) {
        return GAME_TYPES.register(name, supplier);
    }
}

File: src/main/kotlin/qtech/bubbles/init/GameTypes.kt
package qtech.bubbles.init

import qtech.bubbles.BBInternalMod
import qtech.bubbles.common.gametype.AbstractGameMode
import qtech.bubbles.common.init.ObjectInit
import qtech.bubbles.gamemode.ClassicMode
import qtech.bubbles.registry.DeferredRegister
import qtech.bubbles.registry.Registers
import qtech.bubbles.registry.`object`.RegistryObject
import java.util.function.Supplier

/**
 * @author Quinten Jungblut
 * @see AbstractGameMode
 *
 * @since 1.0.0
 */
//@ObjectHolder(addonId = "qbubbles", type = GameType.class)
object GameTypes : ObjectInit<AbstractGameMode?> {
    val GAME_TYPES: DeferredRegister<AbstractGameMode> = DeferredRegister.create(BBInternalMod.ADDON_ID, Registers.GAME_TYPES)
    val CLASSIC_MODE: RegistryObject<ClassicMode> = register("classic") { ClassicMode() }
    private fun <T : AbstractGameMode> register(name: String, supplier: Supplier<T>): RegistryObject<T> {
        return GAME_TYPES.register(name, supplier)
    }
}